bazel/packaging/tool.go

```diff
@@ -27,31 +27,87 @@ import (
 	"time"
 )
 
-type fipsPkgConfig struct {
-	Module string `json:"module"`
-	Config string `json:"config"`
+type fileConfig struct {
+	Path       string `json:"path"`
+	Name       string `json:"name"`
+	SourcePath string `json:"source"`
 }
 
 type pkgConfig struct {
-	RedpandaBinary    string         `json:"redpanda_binary"`
-	RPUtil            *string        `json:"rp_util"`
-	IOTune            *string        `json:"iotune"`
-	HWLocCalc         *string        `json:"hwloc_calc"`
-	HWLocDistrib      *string        `json:"hwloc_distrib"`
-	OpenSSL           *string        `json:"openssl"`
-	RPKBinary         *string        `json:"rpk"`
-	SharedLibraries   []string       `json:"shared_libraries"`
-	DefaultYAMLConfig *string        `json:"default_yaml_config"`
-	Owner             int            `json:"owner"`
-	DirectoryMode     bool           `json:"directory_mode"`
-	Fips              *fipsPkgConfig `json:"fips"`
+	PackageDirectories []string     `json:"package_dirs"`
+	Binaries           []fileConfig `json:"binaries"`
+	PackageFiles       []fileConfig `json:"package_files"`
+	DirectoryMode      bool         `json:"directory_mode"`
+	Owner              int          `json:"owner"`
+}
+
+type dirNode struct {
+	Name     string
+	Children map[string]*dirNode
+}
+
+func splitPathElements(path string) []string {
+	dir, last := filepath.Split(path)
+	if dir == "" {
+		return []string{last}
+	}
+	return append(splitPathElements(filepath.Clean(dir)), last)
+}
+
+func buildDirTree(dirs []string) *dirNode {
+	root := &dirNode{
+		Name:     "",
+		Children: make(map[string]*dirNode),
+	}
+
+	for _, dir := range dirs {
+		current := root
+		pathParts := splitPathElements(dir)
+
+		for _, part := range pathParts {
+			if part == "" {
+				continue
+			}
+			if _, exists := current.Children[part]; !exists {
+				current.Children[part] = &dirNode{
+					Name:     part,
+					Children: make(map[string]*dirNode),
+				}
+			}
+			current = current.Children[part]
+		}
+	}
+
+	return root
+}
+
+func traverseDirTree(node *dirNode, parentPath string, fun func(string) error) error {
+	if node == nil {
+		return nil
+	}
+
+	currentPath := filepath.Join(parentPath, node.Name)
+	if currentPath != "" {
+		if err := fun(currentPath); err != nil {
+			return err
+		}
+	}
+
+	for _, child := range node.Children {
+		if err := traverseDirTree(child, currentPath, fun); err != nil {
+			return err
+		}
+	}
+
+	return nil
+
 }
```

Comment:
Probably simpler to use `MapFS` to build a fake directory structure (you can specify empty directories in the map): https://pkg.go.dev/testing/fstest#MapFS then use fs.WalkDir: https://pkg.go.dev/io/fs#WalkDir

Reply:
TBH this whole program could probably be simpler then.

Basically build up this in memory map as a fs.FS, then you pass that to another function to create a tarball or a directory on a different FS